monai/data/wsi_reader.py

```diff
@@ -0,0 +1,310 @@
+    - The `get_data` call fetches the image data, as well as meta data.
+    - `verify_suffix` verifies
+    - `_verify_output` verifies the extracted patch to be a two dimensional RGB/RGBA image
+
+    The following methods needs to be implemented for any concrete implementation of this class:
+
+    - `_reader` returns a whole slide image reader module that given filename and additional arguments,
+        returns image object.
+    - `_get_size` returns the size of the whole slide image of a given wsi object at a given level.
+    - `_get_level_count` returns the number of levels in the whole slide image
+    - `_get_patch` extracts and returns a patch image form the whole slide image
+    - `_get_metadata` extracts and returns metadata for a whole slide image and a specific patch.
+
+
+    """
+
+    supported_formats: List[str] = []
+
+    def __init__(self, level: int, **kwargs):
+        super().__init__()
+        self.level = level
+        self.kwargs = kwargs
+        self.metadata: Dict[Any, Any] = {}
+
+    @property
+    @abstractmethod
+    def _reader(self):
+        """Returns a whole slide image reader module that given filename and additional arguments, returns image object."""
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def _get_size(self, wsi, level):
+        """Returns the size of the whole slide image of a given wsi object at a given level."""
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def _get_level_count(self, wsi):
+        """Returns the number of levels in the whole slide image."""
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def _get_patch(
+        self, wsi, location: Tuple[int, int], size: Tuple[int, int], level: int, dtype: DtypeLike, mode: str
+    ):
+        """Extracts and returns a patch image form the whole slide image."""
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def _get_metadata(self, wsi, patch: np.ndarray, location: Tuple[int, int], size: Tuple[int, int], level: int):
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    def read(self, data: Union[Sequence[PathLike], PathLike, np.ndarray], **kwargs):
+        """
+        Read image data from given file or list of files.
+
+        Args:
+            data: file name or a list of file names to read.
+            kwargs: additional args for backend reading API in `read()`, will override `self.kwargs` for existing keys.
+                more details in `cuCIM`: https://github.com/rapidsai/cucim/blob/v21.12.00/cpp/include/cucim/cuimage.h#L100.
+
+        Returns:
+            image object or list of image objects
+
+        """
+        wsi_list: List = []
```

Comment:
Is there any use case for a list of WSI files?
I think the `get_data()` API doesn't support that?

Thanks.

Reply:
That's a good point! I will simplify it.

Reply:
Actually we shouldn't check it. This class is inherited from `ImageReader`, and it expect filename or list of filenames, so we should not make this argument more specific. It is violation of Liskov substitution principle.

https://github.com/Project-MONAI/MONAI/blob/443fc0cf11e644d1b85ae2a6a43e7820ecdc604c/monai/data/image_reader.py#L79

Reply:
Hi @drbeh ,

I understand the principle point. The problem is that: the other readers can `read()` a list of images and `get_data()` also supports a list of image objects:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/image_reader.py#L99
but here your `get_data()` function doesn't(or no need) work with a list of image objects:
https://github.com/Project-MONAI/MONAI/pull/4107/files#diff-e0abb01781c5bed5c90cf46759d8e2aeac0793b86bd48ab49b40b6b7f7f5c1c2R94
How about this:
1. Keep your current API and args.
2. Add a check in the `get_data()`: if passing `wsi` object, ok, if passing a list of `wsi` objects, should only contains one item in it and use it. For example:
```py
def get_data(self, wsi, ...):
    if isinstance(wsi, (tuple, list)):
        if len(wsi) != 1:
            raise ValueError("only support 1 `wsi` image object so far.")
        wsi = wsi[0]
```

What do you think?

Thanks.

Reply:
Hi @Nic-Ma, thanks. That's a good idea but I can make them to load the same patch from multiple wsi objects. I think this way would be more compatible with the rest of readers.

Reply:
Ah, the output of `get_data` is also too specific! I'll go with input checking.